add list runs command

hamelsmu opened the pull request:
~TODO: I need to add tests~

@valayDave @savingoyal this is ready for review

This will close #602

romain-intel commented on metaflow/plugins/list_cli.py:
```diff
@@ -0,0 +1,69 @@
+@click.option(
+    "--file",
+    default=None,
+    help="Save the run list to file as json.",
```
The code makes it look like if you don't specify anything it doesn't print as json nor does it output to a file? There doesn't seem to also respect the quiet flag.

hamelsmu replied:
Great catch, I accidentally forgot to add that back in while refactoring!  I will fix this!

hamelsmu replied:
Ok I fixed this

romain-intel commented on metaflow/plugins/list_cli.py:
```diff
@@ -0,0 +1,69 @@
+def runs(ctx, num_runs, as_json, file):
+    run_list = _fetch_runs(ctx.obj.flow.name, num_runs)
+    if not run_list:
+        ctx.obj.echo("No runs found for flow: {name}".format(name=ctx.obj.flow.name))
```
Nit: this message is incorrect (potentially) if num_runs <= 0

hamelsmu replied:
I thought [empty lists are falsy](https://docs.python.org/3/library/stdtypes.html#truth-value-testing)?  I have a feeling I'm missing something else, just trying to learn, thanks.

hamelsmu replied:
Ohh you mean if the cli argument itself is <= 0 !!!! I didn't think about this.  Okay

hamelsmu replied:
ok I fixed this via https://github.com/Netflix/metaflow/pull/1037/commits/fbba6be401b9f4a50f3accaabcf20b85dad4d443

romain-intel reviewed (commented):
I don't have a huge issue with this PR but it does seem somewhat of a point solution and maybe we should think about the "list" API a bit more generally. It may be that we want to add more types of "lists" after (not saying to do it now but at least to think about it a bit).

hamelsmu commented:
Oh I need to add the bit about the `--quiet` flag

savingoyal commented on metaflow/plugins/list_cli.py:
```diff
@@ -0,0 +1,77 @@
+)
+@list.command(help="List recent runs for your flow.")
+@click.pass_context
+def runs(ctx, num_runs, as_json, file):
```
Can you also add `--user` and `--my-runs` flags as well? - take a look at `batch_cli.py`

hamelsmu replied:
:thumbs_up:

hamelsmu replied:
Thanks I implemented this in https://github.com/Netflix/metaflow/pull/1037/commits/e0682b15a576989cfd408db7708e26870bc282ed

savingoyal commented on metaflow/plugins/list_cli.py:
```diff
@@ -0,0 +1,77 @@
+    counter = 1
+    try:
+        flow = Flow(flow_name)
+    except MetaflowNotFound:
```
We need to set the correct namespace here to ensure that the `Flow` or `Run` can be found.

hamelsmu replied:
@romain-intel perhaps you can shed some light here as I think I am confused b/w what Savin is saying and what you are saying elsewhere?  

Happy to have a sync conversation as well around this.  Thank you both for your help

hamelsmu commented:
@savingoyal I was discussin this with @valayDave and some questions:

1. Do we want to add a `--namespace` flag as well?
2. If we want to add 1, should we not have the `--user` flag and let people pass the namespace `user:<username>` instead?
3. If we want to keep both a `--namepsace` and `--user` flag, should they be mutually exclusive, or how do we want to handle collisions in this case, should we throw an error if there if there is a `--namespace` argument with `user:<username>` etc?  

Thanks

savingoyal commented:
How about we add both? It's usually a lot easier as a user for me to do `python flow.py list --my runs` or `python flow.py list --user=hamel runs`. We can also support `namespace` and make the set - `my`, `user` and `namespace` mutually exclusive. wdyt?

valayDave commented:
Ok. Here are some key questions that would need decisions to quickly converge on code changes.
1. What are the levels of filters were wish to support. Some I can think of are: `tags`, `namespace`, `user`, `date`, `branch`/`project`, `success`/`failure`.  
2. When options for these filters are provided from CLI, then will the resultant runs be filtered based on the union of filters or will there be some compatibility Constraints. An example constraint: You can provide `date` or `tags` but not both. Not saying we need to place this constraint. Just giving an example.  
3. Should the namespace command be `list` or something else ? Since we right now have something like `card list` or `batch list-jobs` etc. Similarly why not `runs list`. I agree that it is quite close to `run` making it a little confusing for someone at some point. 
4. Given the convergence on the above decisions, what is the final command spec. (Example command with all arguments)

hamelsmu commented:
@savingoyal I have added the `--namespace` flag so that we have something to look at and consider, but I think we should definitely discuss @valayDave 's [comments above](https://github.com/Netflix/metaflow/pull/1037#issuecomment-1122977965).   Please LMK your thoughts

romain-intel commented:
The namespace is implicitly the user (which I thought was the initial intent) but yes, this is exactly the type of conversation I was mentioning we should have around this more general "listing" API. These types of higher-level options (independent in effect of whether you are listing runs, steps, artifacts, etc) is useful.

romain-intel commented on metaflow/plugins/list_cli.py:
```diff
@@ -0,0 +1,101 @@
+    if my_runs:
+        namespace = "user:{}".format(util.get_username())
+    if user:
+        namespace = "user:{}".format(user)
```
default should probably be my-runs (which is what it was before).

hamelsmu replied:
@romain-intel Just to clarify, I think what you are saying is that because of the call to `namespace(namespace)` in `cli.py` makes the `--my-runs` flags necessary?  Is my understanding correct? 

cc: @savingoyal incase I misunderstood your instructions earlier to add this flag as well

romain-intel replied:
no, I meant that right now, if you don't specify `my-runs` or `user`, it will do `namespace(None)` (or at least you should do that -- see my comment below) which means "no namespace filter". I am saying that if there is nothing specified, we should default to the behavior of `my-runs` because that is probably the most natural thing. We can still have an option (to be explicit). Of course that does bring the "how do you list all runs across namespaces" but I still feel that having the `my-runs` be the "default" behavior is probably more natural. At least it is the one that best answers the user story that @valayDave had in the original issue.

hamelsmu replied:
ok I resolved this [with these changes](https://github.com/Netflix/metaflow/pull/1037#issuecomment-1123825173)

romain-intel commented on metaflow/plugins/list_cli.py:
```diff
@@ -0,0 +1,101 @@
+        for run in Flow(flow_name).runs():
+            if counter > num_runs:
+                break
+            tags = [t for t in run.tags]
```
you don't need this. the call to `namespace(namespace)` should take care of filtering for you (it was already filtered by username by default).

hamelsmu replied:
Gotcha, ok I changed it to setting the namespace before retrieving runs, instead of manually filtering tags [as described here](https://github.com/Netflix/metaflow/pull/1037#issuecomment-1123825173)

hamelsmu commented:
ok @romain-intel I made the changes (If I understood correctly) [in this commit](https://github.com/Netflix/metaflow/pull/1037/commits/742b080a7b45f6be699c1539ba215ca881d51604)

Please let me know if I am understanding correctly, hopefully, I'm getting closer!

hamelsmu commented:
ok @romain-intel I made the changes (If I understood correctly) [in this commit](https://github.com/Netflix/metaflow/pull/1037/commits/742b080a7b45f6be699c1539ba215ca881d51604)

I made the following changes:

1. removed the `--my-runs` flag and made that the default behavior
2. added a `--show-all` flag that calls `namespace(None)` to allow you to access the global namespace
3. Filtered runs by calling `namespace(<namespace>)`, instead of iterating through tags of each run

Please let me know if I am understanding correctly, hopefully, I'm getting closer!

We still have to have a discussion of this interface in general.    Note: this PR was also meant as a "Good First Issue" for me tackle with regards to learning the code base of Metaflow (just to let you know the dual purpose of this PR)

hamelsmu commented:
Feedback from @savingoyal 

> My take is that we can expand list in the future to include cards etc. and supporting just --namespace (and maybe --my and --user as well) for now is a good start. We can always introduce more flags as we learn usage patterns.  --namespace already takes care of tags , user , date , project etc.

One minor TODO for @hamelsmu requested by @savingoyal 

> can we do list --all runs instead of list --show-all runs

savingoyal commented:
Converted the PR to draft since it is not ready for review yet.
